src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -1291,6 +1124,15 @@ public void save() throws IOException {
         synchronized (this) {
             PipelineIOUtils.writeByXStream(this, loc, XSTREAM2, isAtomic);
             SaveableListener.fireOnChange(this, file);
+            if (Boolean.TRUE.equals(completed)) {
```

Comment:
@jglick No! 🐛 🐛 🐛 and a whole nest of 🐛s.   This is not the right place to be terminating the AsynchronousExecution - if an IOException never occurs, the build never finishes. 

You're welcome to do an extra on-save check _somewhere logical_ -- because we have had hard-to-trace issues with the build not getting killed before, but this makes undead Pipelines more likely.

Reply:
There is explanation elsewhere of why we _have_ to do the termination here. I can think about whether it is possible to do so more forcefully, such as in a `finally` block.

Reply:
> if an IOException never occurs

I suppose you mean if an `IOException` _does_ occur.

> the build never finishes

As noted in a separate thread, even if no subsequent call to `save` is ever made, it is misleading to say that the build is “not finished”. Rather, there would be a stale `AsynchronousExecution`, which likely has little or no visible impact.

Reply:
(see https://github.com/jenkinsci/workflow-job-plugin/pull/27#discussion_r221294107 for fuller discussion)

Reply:
A stale AsynchronousExecution translates to essentially a "zombie" job that proves highly irritating to users -- it seems like it never completes even if the build is marked as done.  Also it holds onto the OneShotExecutor indefinitely.

Reply:
Again the `AsynchronousExecution` normally has no visible effect on anything—it only appears while we are actively running Groovy code, which of course we are not if the build is `completed`.

And again if some problem in _existing_ code (whether in `workflow-job` or `workflow-cps`) results in a condition whereby `<completed>true</completed>` never gets written to `build.xml`, the `AsynchronousExecution` is the least of your concerns, since after the next restart `onLoad` will be confused. Possibly #93 will defend against that case—figure out that `completed` is wrong and save the correction—but at that point any old `Executor`s are no longer an issue anyway (and there should be no new `AfterRestartTask`).

Reply:
I think the after-restart case should have been fixed by #104.